src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
 			break;
 			}
 
-	// Ideally, we would always call select on the fds to see which
-	// are ready, and return the soonest. Unfortunately, that'd mean
-	// one select-call per packet, which we can't afford in high-volume
-	// environments.  Thus, we call select only every SELECT_FREQUENCY
-	// call (or if all sources report that they are dry).
+	// If there aren't any sources and exit_only_after_terminate is false, just
+	// return an empty set of sources. We want the main loop to end.
+	if ( Size() == 0 && ( ! BifConst::exit_only_after_terminate || terminating ) )
+		return ready;
 
-	++call_count;
+	double timeout = -1;
+	IOSource* timeout_src = nullptr;
+	timer_expired = false;
 
-	IOSource* soonest_src = 0;
-	double soonest_ts = 1e20;
-	double soonest_local_network_time = 1e20;
-	bool all_idle = true;
+	PktSrc* pkt_src = pkt_srcs.empty() ? nullptr : pkt_srcs.front();
 
-	// Find soonest source of those which tell us they have something to
-	// process.
-	for ( SourceList::iterator i = sources.begin(); i != sources.end(); ++i )
+	// If we have an active source of data then find the next timeout value from
+	// those sources. Start by getting the next timeout from the timer manager
+	// since it's not an IOSource.
+	// TODO: how do we handle multiple timer managers here?
+	if ( pkt_src || broker_mgr->Active() )
 		{
-		if ( ! (*i)->src->IsIdle() )
+		timeout = timer_mgr->GetNextTimeout();
+
+		for ( auto src : sources )
 			{
-			all_idle = false;
-			double local_network_time = 0;
-			double ts = (*i)->src->NextTimestamp(&local_network_time);
-			if ( ts >= 0 && ts < soonest_ts )
+			if ( src->src->IsOpen() )
 				{
-				soonest_ts = ts;
-				soonest_src = (*i)->src;
-				soonest_local_network_time =
-					local_network_time ?
-						local_network_time : ts;
+				double next = src->src->GetNextTimeout();
+				if ( timeout == -1 || ( next >= 0.0 && next < timeout ) )
+					{
+					timeout = next;
+					timeout_src = src->src;
+					}
 				}
 			}
 		}
 
-	// If we found one and aren't going to select this time,
-	// return it.
-	int maxx = 0;
+	// If the active packet source isn't live or the packet source is ready (i.e.
+	// we're in pseudo-realtime mode and it's time for a packet), always insert
+	// the packet source into the list of ready sources.
+	if ( pkt_src && pkt_src->IsOpen() && ! pkt_src->IsLive() && pkt_src->GetNextTimeout() == 0 )
```

Comment:
Similar question as above: Would be nice if we didn't need this special-casing here. Could the packet source do the right things through its IOSource API implementation so that it lands in ready automatically when necessary? 

(It's ok if the answer is "no" and we do need the special-case here; just checking if that's indeed the case, as it would make for nice API if the mgr wouldn't event need to know at all that a source is a pkt src).

Reply:
There was an issue when I was originally working on the kqueue() code where it would get stuck in the middle of processing a file and never return from the call to `kevent`. I think that was before I ended up just inserting a timeout of zero for trace files, so it looks like that issue has done away now. I just removed this check, reran the tests, and everything passed so I'll just take it out.

Reply:
This does have the same problem with kqueue still. It'll get stuck in the middle of processing a file and ends up spinning on a zero timeout never adding the packet source to the set of ready ones. The special case here was basically "packet files are always ready and we don't need to check them."

Reply:
This one is compounded by a bug in the if/else block in `net_run`. Basically, if there's nothing in the ready list and we expire a timer in FindReadySources() in not-live/not-PR mode, then we never process anything. No timers, no packets, nothing. Then we jump back into FindReadySources() and end up looping forever.

Reply:
Fixed/removed in the changes for TimerMgr.